CLI Manager

miniche opened the pull request:
CentralReport is designed to be easier as possible and friendly. The configuration is stored in a file (/etc/centralreport/centralreport.cfg) and can be edited with a text editor.

This pull request adds a "CLI Manager". This manager is a friendly way to configure and to manage your CentralReport software. The Manager will be able to :
- [x] Check the daemon status and the processus ID (PID)
- [x] Start or stop the daemon
- [x] Edit the web server port for the "Standalone app"
- [ ] ~~Edit the account token (CentralReport Online)~~

This Manager uses the **Urwid library** (http://excess.org/urwid/). This library is very powerful and so easy. In the future, we can imagine lots of new functions with this library (get stats in the CLI like htop? configure more options?).

The main screen:
![1](https://f.cloud.github.com/assets/1638158/983388/804b0648-0858-11e3-815c-b27e77b7bc52.png)

The manager accepts the new "wizard" option. This will be used during the installation, and displays two screens:
- A welcome screen like this:
  ![wizard 1](https://f.cloud.github.com/assets/1638158/851289/6d9cd7a2-f498-11e2-8d8d-cc861d8343b6.png)
- After, ask for the internal web server activation:
  ![capture decran 2013-07-23 a 23 26 56](https://f.cloud.github.com/assets/1638158/851295/80bf9e6e-f498-11e2-830e-2e6a582ade9e.png)

Some details:
- The manager must be executed as root in order to start/restart/stop the CR daemon and to write the new configuration file
- The configuration file is written automatically just before the manager ends (on selecting the "Save and Quit" button or with the CTRL+C)
- The manager requires that CentralReport is currently installed on the current host because it uses the binary script (/usr/local/bin/centralreport) to control the daemon. Moreover, we are sure that the folder which contains the configuration file (/etc/centralreport/) already exists.
- The manager is able to handle the POSIX signals, like SIGINT to end properly (CTRL+C)

In the first version, the manager will be able to:
- Check whether CentralReport is running and displays its PID number
- Start/Restart/Stop the CentralReport daemon
- Modify the internal web server configuration (activation and port number), called "Standalone app"

miniche committed:
Added a CLI Manager using the Urwid library

miniche committed:
Added docstrings in the CLI Manager functions

Ninir commented on centralreport/cr/cli.py:
```diff
@@ -0,0 +1,178 @@
+
+"""
+    CentralReport - CLI module (WIP)
+        Manages the common line interface
```
`Common`? isn't it `command`? :smile:

Ninir commented on centralreport/cr/cli.py:
```diff
@@ -0,0 +1,178 @@
+
+    def input_handle(self, input):
+        """
+            Default behavior when the user press a key
```
I would say `presses`

Ninir commented on centralreport/cr/cli.py:
```diff
@@ -0,0 +1,178 @@
+    """
+        Initializes the screen with a header (3 lines) and a footer (1 line).
+        The default background color is blue for the bars and black for the content.
+        This functions displays nothing, see display_screen().
```
`This function` without a `s`
Instead of `displays nothing`, I would say `This function does not display anything`, seems less weird 
(As in french with "Cette fonctionne affiche rien", remplacée par "Cette fonction n'affiche rien")

Ninir commented on centralreport/cr/cli.py:
```diff
@@ -0,0 +1,178 @@
+
+def display_screen(input_handle):
+    """
+        This functions sends the content to the screen.
```
This function (let's do it everywhere instead of receiving 50 notifications ^^)

Ninir commented on centralreport/cr/cli.py:
```diff
@@ -0,0 +1,178 @@
+    """
+        This functions sends the content to the screen.
+
+        @param input_handle: Callback functions called when the user press a key
```
presses

Ninir commented on centralreport/cr/cli.py:
```diff
@@ -0,0 +1,178 @@
+
+def create_radio_item(group, name, callback):
+    """
+        Create a radio item, in a group. See urwid.RadioButton for more details
```
Creates

Ninir commented on centralreport/manager.py:
```diff
@@ -0,0 +1,194 @@
+
+"""
+    CentralReport - Manager
+        This script displays the CLI Manager, used to modify the CentralReport configuration.
```
What about `This CLI manager is used to modify the CentralReport configuration`?

Ninir commented on centralreport/manager.py:
```diff
@@ -0,0 +1,194 @@
+    CentralReport - Manager
+        This script displays the CLI Manager, used to modify the CentralReport configuration.
+
+
```
Additional empty line

Ninir commented on centralreport/manager.py:
```diff
@@ -0,0 +1,194 @@
+        cr.cli.WindowCli.__init__(self)
+
+        self.title = 'Online agent configuration'
+        self.subtitle = 'CentralReport Online is the best choose to monitor your host ' \
```
Do you mean `choice`?

miniche committed:
The Urwid library is now added to the sys.path in the main centralreport.py module

miniche committed:
Added a "DialogBox" object in the CLI module

Used to interact with the user, displaying texts or errors.

miniche committed:
Linked "Config" class to the CLI Manager

miniche committed:
Improved Dialog object in the "cli" module

miniche committed:
Improved CLI Manager with major improvements

- The manager must be run as root to perform some tasks (write new configuration, restart the daemon, …)
- CentralReport must be already installed on the current host to run the CLI Manager
- The configuration for the internal web server is now written in the configuration file
- The daemon restart when the user exit the CLI Manager to use the new configuration

PS: We must work on the "feature-webservices" branch to finish this PR for the "Online configuration" part.

miniche committed:
[WIP] Working on CLI Manager (MainCli, palette colors, Update the configuration, …)

miniche commented:
I'm working on the "Main" screen:
![cli main](https://f.cloud.github.com/assets/1638158/842413/f5897868-f3a9-11e2-8e1b-f7d304b7a2f9.png)

New stuff tonight!

Guimove commented:
Very good job. I love what is growing in this project. :clap:

Guimove commented on centralreport/manager.py:
```diff
@@ -0,0 +1,296 @@
+    https://github.com/CentralReport
+"""
+
+# The centralreport module is imported first to initialize third-party libraries
```
firstly imported...  ?

Guimove commented on centralreport/manager.py:
```diff
@@ -0,0 +1,296 @@
+        self.subtitle = 'CentralReport always keeps an eye on your system. Receive alerts in real time, ' \
+                        'follow up statistics evolution and much more. \n' \
+                        'The project is open sourced and available at github.com/CentralReport. \n' \
+                        'All the documentation can be found at docs.centralreport.net'
```
documentations ?

Ninir replied:
Nein ^^

miniche committed:
Improved CLI screens

miniche commented:
Some details after the previous commits:
- The manager must be executed as root in order to start/restart/stop the CR daemon and to write the new configuration file
- The configuration file is written automatically just before the manager ends (on selecting the "Save and Quit" button or with the CTRL+C)
- The manager requires that CentralReport is currently installed on the current host because it uses the binary script (/usr/local/bin/centralreport) to control the daemon. Moreover, we are sure that the folder which contains the configuration file (/etc/centralreport/) already exists.
- The manager is able to handle the POSIX signals, like SIGINT to end properly (CTRL+C)

In the first version, the manager will be able to:
- Check whether CentralReport is running and displays its PID number
- Start/Restart/Stop the CentralReport daemon
- Modify the internal web server configuration (activation and port number), called "Standalone app"
- Modify the link to the Online server (activation and user token)

The manager accepts the new "wizard" option. This will be used during the installation, and displays three screens:
- A welcome screen like this:
  ![wizard 1](https://f.cloud.github.com/assets/1638158/851289/6d9cd7a2-f498-11e2-8d8d-cc861d8343b6.png)
- After, ask for the internal web server activation:
  ![capture decran 2013-07-23 a 23 26 56](https://f.cloud.github.com/assets/1638158/851295/80bf9e6e-f498-11e2-830e-2e6a582ade9e.png)
- End by the Online configuration:
  ![capture decran 2013-07-23 a 23 27 01](https://f.cloud.github.com/assets/1638158/851299/899a214e-f498-11e2-9cbe-4e40d3064177.png)

## Waiting for...

I'm waiting for the "feature-webservices" branch to be merged in "develop" to end the "Online configuration" part.

## Remains to be done
- Refresh the main screen when the configuration is modified
- Update the binary script to start the new "manager" with the command "centralreport manager"
- Update the installation scripts to start the wizard just before the end of the procedure
- Update the documentation (docs.centralreport.net)
- Update the pull request description

miniche committed:
Major improvements in the CLI Manager

- The port for the internal web server is now checked when the user choose one new
- The main screen is now refreshed when the user modifies the configuration
- The daemon status ("CentralReport is not running" / "CentralReport is running with PID xxx") is now updated when the user starts/stops/restarts the daemon
- The daemon status is modified when the daemon is starting/stopping ("Starting CentralReport…" / "Stopping CentralReport…" / "Restarting CentralReport…")

Tech:
- The urwid main_loop is now located in the "WindowCli" class. It's more logical and can be updated directly within the object.

miniche committed:
Added "manage" option to the binary script for the CLI Manager

Now, we can start the CLI Manager with the command "centralreport manage" (or "/usr/local/bin/centralreport manage").

miniche committed:
The CLI wizard is now started at the end of the installation

miniche commented:
Hey guys!
Can you play with the manager and give me your feedback and your ideas? :)

To use the manager, CentralReport must be installed on your host.
You can start it with the command "centralreport manage" or "sudo python /usr/local/lib/centralreport/manager.py".
Please note that the "Online" configuration is not available yet and your choice is not saved for this option.

And... a global review is welcome!

miniche committed:
Removed all references to CentralReport Online

This part will be added in a future release.

miniche committed:
In the binary file, the "manager" parameter is now used to start… the CLI Manager.

(Was "manage" before)

miniche commented:
I have removed all references to CentralReport Online in 597aa31 in order to release an intermediate version (0.3.0). The online part will be added in a next release.

According to me, it's all good. You can review this PR :)

miniche committed:
Updated NEWS file for #82 (CLI Manager)

miniche commented:
(PS: I've updated the release notes for the 0.3.0, adding this PR: http://docs.centralreport.net/unix-agent/releases/0.3.0)

Ninir commented on centralreport/cr/cli.py:
```diff
@@ -0,0 +1,196 @@
+    main = urwid.ListBox(content_list)
+
+
+def generate_blank_characters(number):
```
We have to find libraries handling all these functions as there might be better implementations. If not, let's create a different package including our functions to provide more open-source stuff :)

miniche replied:
True. It's the same situation for the modules in the 'utils' package, and most of bash functions.
I think we could look up to improve these parts when we will have more time :)

Ninir commented:
Seems good looking at the code. TIme to review the UI :)

miniche commented:
If you want to check the interface, CentralReport must be installed on your host (please see: https://github.com/CentralReport/CentralReport/pull/82#issuecomment-21827227).

I hope you like it :)

miniche commented:
@Ninir, are you sleeping? :smile: 

It seems that we have a conflict with the develop branch. I will look at it tonight.

miniche committed:
Merge remote-tracking branch 'origin/develop' into feature-manager

Conflicts:
	NEWS

Ninir commented:
I'm not anymore :P

Ninir commented:
Will review asap :)

Ninir commented on NEWS:
```diff
@@ -25,6 +25,15 @@ Version Alpha 0.3.0 (Not released yet)
 
+    - Feature #82: New CLI Manager
+        * A friendly way to modify the CentralReport configuration with
+          interactive menus and checkbox in the user terminal. No need
```
and checkboxes*

Ninir commented on NEWS:
```diff
@@ -25,6 +25,15 @@ Version Alpha 0.3.0 (Not released yet)
+    - Feature #82: New CLI Manager
+        * A friendly way to modify the CentralReport configuration with
+          interactive menus and checkbox in the user terminal. No need
+          to modify manually the configuration file!
```
the configuration file manually!*

Ninir commented on NEWS:
```diff
@@ -25,6 +25,15 @@ Version Alpha 0.3.0 (Not released yet)
+          interactive menus and checkbox in the user terminal. No need
+          to modify manually the configuration file!
+          Can be used with the command "centralreport manager".
+        * Called during the installation to ask to the user the basic
```
to ask the user*

Ninir commented on NEWS:
```diff
@@ -25,6 +25,15 @@ Version Alpha 0.3.0 (Not released yet)
+          Can be used with the command "centralreport manager".
+        * Called during the installation to ask to the user the basic
+          configuration, like the port for the interal web server. Not
+          started when the user use the '-s' parameter (silent installation).
```
I would go for `Not started when using the '-s' parameter...`

Ninir commented on bash/functions.inc.sh:
```diff
@@ -65,6 +65,21 @@ function stop_cr(){
+#
+function start_wizard(){
+
+    if [ "${ARG_S}" != true ]; then
```
Isn't it possible to omit the `!= true` ?

as in `if [ "${ARG_S}" ]; then`

miniche replied:
`if [ ! "${ARG_S}" ]; then` will work, yes (you have just omited the exclamation mark).

Ninir commented on centralreport/manager.py:
```diff
@@ -0,0 +1,280 @@
+
+    def validate(self, button):
+        """
+            Triggered when the user press the "OK" button
```
presses*

Ninir commented on centralreport/manager.py:
```diff
@@ -0,0 +1,280 @@
+                        'The project is open sourced and available at github.com/CentralReport. \n' \
+                        'All the documentation can be found at docs.centralreport.net'
+
+        caption = 'This wizard helps you to configure main options. You will able to update the ' \
```
You will be*

Ninir commented on centralreport/manager.py:
```diff
@@ -0,0 +1,280 @@
+
+    def input_handle(self, input):
+        """
+            Default behavior when the user press a key
```
presses*, and same thing line 237

miniche committed:
Minor spelling fixes

miniche commented:
So... Can I merge? :)

Ninir commented:
Yup :P
